test/run-drun/map-upgrades/map0.mo

```diff
@@ -0,0 +1,82 @@
+import Prim "mo:⛔";
+import Cycles = "../cycles/cycles";
+import Lib "node0";
+
+// A naive, distributed map from Nat to Text.
+// Illustrates dynamic installation of imported actor classes.
+// Uses a fixed number of nodes, dynamically installed on demand 
+// .. and upgraded with a call to upgradeNodes() (without data loss)
+
+actor a {
+
+  type Key = Nat;
+  type Value = Text;
+
+  // Number of Nodes
+  let n = 8;
+
+  type Node = Lib.Node;
+
+  stable let savedNodes : [var ?(actor{})] = Prim.Array_init(n, null);
+
+  let nodes : [var ?Node] = Prim.Array_init(n, null);
+
+  // Would be nice if these were both tail calls on the platform
+  public func lookup(k : Key) : async ?Value {
+    switch (nodes[k % n]) {
+      case null null;
+      case (?node) await node.lookup(k);
+    };
+  };
+
+  public func insert(k : Key, v : Value) : async () {
+    let i = k % n;
+    let node = switch (nodes[i]) {
+      case null {
+        Cycles.add(2_000_000_000_000);
+        let n = await Lib.Node(i); // dynamically install a new Node
+        nodes[i] := ?n;
+        n;
+      };
+      case (?node) node;
+    };
+    await node.insert(k, v);
+  };
+
+  system func preupgrade () {
+     for (i in nodes.keys()) {
+       savedNodes[i] := nodes[i];
+     }
+  };
+
+  public func upgradeNodes() {
```

Comment:
Is it better to put this in `postupgrade`?

Reply:
You can't because `postupgrade` can't send messages - you need to design a user-level protocol to do upgrades of instances, unfortunately.